File: android/core/calc/src/test/kotlin/com/stocktracker/core/calc/RowDerivationTest.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.ManualPriceEntry
import com.stocktracker.core.model.PortfolioRow
import com.stocktracker.core.model.Position
import com.stocktracker.core.model.PositionType
import com.stocktracker.core.model.Quote
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class RowDerivationTest {

    /** Same-currency identity conversion — keeps the arithmetic legible in these tests. */
    private val identity: Convert = { amount, _, _ -> amount }

    private fun openLot(qty: Double, buyPrice: Double, buyDate: String) = Position(
        id = "1", ticker = "AAPL", name = "Apple", type = PositionType.STOCK,
        quantity = qty, buyPrice = buyPrice, buyDate = buyDate, currency = "USD",
    )

    @Test
    fun `open lot with live quote uses quote price`() {
        val lot = openLot(qty = 10.0, buyPrice = 100.0, buyDate = "2024-01-01")
        val quote = Quote(ticker = "AAPL", price = 120.0, change = 2.0, changePercent = 1.6, currency = "USD", name = "Apple", lastUpdated = "2024-06-01")

        val row = deriveRow(
            lots = listOf(lot), quoteRaw = quote, loadingRaw = false, errorRaw = null, manualRaw = null,
            dividends = emptyList(), today = "2024-06-01", convert = identity,
        )

        assertFalse(row.isClosed)
        assertEquals(120.0, row.currentPrice, 1e-9)
        assertEquals(1200.0, row.currentValue, 1e-9)
        assertEquals(1000.0, row.costBasis, 1e-9)
        assertEquals(200.0, row.pnl, 1e-9)
        assertEquals(20.0, row.pnlPercent, 1e-9)
        assertEquals(20.0, row.dailyChange, 1e-9)
        assertFalse(row.priceIsManual)
    }

    @Test
    fun `closed lot values at average sell price with zero current value`() {
        val lot = openLot(qty = 10.0, buyPrice = 100.0, buyDate = "2024-01-01").copy(sellPrice = 150.0, sellDate = "2024-03-01")

        val row = deriveRow(
            lots = listOf(lot), quoteRaw = null, loadingRaw = false, errorRaw = null, manualRaw = null,
            dividends = emptyList(), today = "2024-06-01", convert = identity,
        )

        assertTrue(row.isClosed)
        assertEquals(150.0, row.currentPrice, 1e-9)
        assertEquals(0.0, row.currentValue, 1e-9)
        assertEquals(500.0, row.pnl, 1e-9)
        assertEquals(50.0, row.pnlPercent, 1e-9)
    }

    @Test
    fun `manual price used when no quote is available`() {
        val lot = openLot(qty = 5.0, buyPrice = 50.0, buyDate = "2024-01-01")
        val manual = ManualPriceEntry(price = 60.0, updatedAt = "2024-05-01")

        val row = deriveRow(
            lots = listOf(lot), quoteRaw = null, loadingRaw = false, errorRaw = null, manualRaw = manual,
            dividends = emptyList(), today = "2024-06-01", convert = identity,
        )

        assertTrue(row.priceIsManual)
        assertEquals(60.0, row.currentPrice, 1e-9)
        assertEquals(300.0, row.currentValue, 1e-9)
        assertEquals(50.0, row.pnl, 1e-9)
        assertEquals("2024-05-01", row.manualPriceDate)
    }

    @Test
    fun `no quote and no manual falls back to break-even at avg buy price`() {
        val lot = openLot(qty = 5.0, buyPrice = 50.0, buyDate = "2024-01-01")

        val row = deriveRow(
            lots = listOf(lot), quoteRaw = null, loadingRaw = false, errorRaw = null, manualRaw = null,
            dividends = emptyList(), today = "2024-06-01", convert = identity,
        )

        assertEquals(50.0, row.currentPrice, 1e-9)
        assertEquals(0.0, row.pnl, 1e-9)
        assertFalse(row.priceIsManual)
        assertNull(row.irr) // no usable price source — must not guess
    }

    @Test
    fun `portfolio irr matches single-position xirr`() {
        val position = openLot(qty = 1.0, buyPrice = 1000.0, buyDate = "2020-01-01")
        val row = PortfolioRow(
            ids = listOf(position.id), ticker = "AAPL", name = "Apple", type = PositionType.STOCK,
            currency = "USD", nativeCurrency = "USD", lots = 1, positions = listOf(position),
            totalQuantity = 1.0, avgBuyPrice = 1000.0, firstBuyDate = "2020-01-01",
            currentPrice = 1100.0, currentValue = 1100.0, costBasis = 1000.0, pnl = 100.0,
            pnlPercent = 10.0, dividendIncome = 0.0, totalReturn = 100.0, loading = false,
            priceIsManual = false, irr = 0.10, isClosed = false, dailyChange = 0.0,
        )

        val irr = computePortfolioIrr(
            positions = listOf(position), rows = listOf(row), dividendsByTicker = emptyMap(),
            displayCurrency = "USD", today = "2021-01-01", convert = identity,
        )

        assertEquals(0.10, irr!!, 0.01)
    }

    @Test
    fun `portfolio irr is null while any row is still loading`() {
        val position = openLot(qty = 1.0, buyPrice = 1000.0, buyDate = "2020-01-01")
        val row = PortfolioRow(
            ids = listOf(position.id), ticker = "AAPL", name = "Apple", type = PositionType.STOCK,
            currency = "USD", nativeCurrency = "USD", lots = 1, positions = listOf(position),
            totalQuantity = 1.0, avgBuyPrice = 1000.0, firstBuyDate = "2020-01-01",
            currentPrice = 1000.0, currentValue = 1000.0, costBasis = 1000.0, pnl = 0.0,
            pnlPercent = 0.0, dividendIncome = 0.0, totalReturn = 0.0, loading = true,
            priceIsManual = false, irr = null, isClosed = false, dailyChange = 0.0,
        )

        val irr = computePortfolioIrr(
            positions = listOf(position), rows = listOf(row), dividendsByTicker = emptyMap(),
            displayCurrency = "USD", today = "2021-01-01", convert = identity,
        )

        assertNull(irr)
    }
}

File: src/utils/rowDerivation.test.ts
import { describe, expect, it } from 'vitest'
import type { Position, Quote } from '../types'
import { computePortfolioIrr, deriveRow, type DeriveRowInput } from './rowDerivation'

const convert = (a: number, from: string, to: string) => {
  const r: Record<string, number> = { CZK: 1, USD: 20 }
  return from === to ? a : (a * r[from]) / r[to]
}
const lot = (o: Partial<Position>): Position => ({
  id: 'l', ticker: 'T', name: 'T', type: 'stock', quantity: 10, buyPrice: 100, buyDate: '2024-01-01', currency: 'USD', ...o,
})
const quote: Quote = { ticker: 'T', price: 120, change: 2, changePercent: 1.7, currency: 'USD', name: 'T', lastUpdated: '' }
const base = (lots: Position[], extra: Partial<DeriveRowInput> = {}) =>
  deriveRow({ lots, loading: false, error: null, dividends: [], taxOverrides: {}, today: '2026-09-25', convert, displayCurrency: 'CZK', ...extra })

describe('deriveRow', () => {
  it('values an open lot at the quote', () => {
    const r = base([lot({})], { quote })
    expect(r.currentValue).toBe(1200)
    expect(r.pnl).toBe(200)
    expect(r.isClosed).toBe(false)
  })

  it('treats a lot sold at price 0 as closed, not also open', () => {
    const r = base([lot({ sellPrice: 0, sellDate: '2025-01-01' })])
    expect(r.isClosed).toBe(true)
    expect(r.currentValue).toBe(0)
    expect(r.pnl).toBe(-1000)
  })

  it('keeps averages finite for a zero-quantity lot', () => {
    const r = base([lot({ quantity: 0 })], { quote })
    expect(Number.isFinite(r.avgBuyPrice)).toBe(true)
    expect(Number.isFinite(r.currentPrice)).toBe(true)
  })
})

describe('computePortfolioIrr', () => {
  it('is not blanked by a closed row whose own IRR has no solution', () => {
    const open = base([lot({ id: 'o', ticker: 'O' })], { quote: { ...quote, ticker: 'O' } })
    const lost = base([lot({ id: 'x', ticker: 'X', sellPrice: 0, sellDate: '2025-01-01' })])
    expect(lost.irr).toBeNull()
    const irr = computePortfolioIrr([...open.positions, ...lost.positions], [open, lost], new Map(), {}, 'CZK', '2026-09-25', convert)
    expect(irr).not.toBeNull()
  })
})
